base/multidimensional.jl

```diff
@@ -619,91 +648,40 @@ getindex{T<:Real}(B::BitArray, I0::UnitRange{T}) = getindex(B, to_index(I0))
         end
 
         storeind = 1
+        Xc, Bc = X.chunks, B.chunks
         @nloops($N, i, d->I_d,
                 d->nothing, # PRE
                 d->(ind += stride_lst_d - gap_lst_d), # POST
                 begin # BODY
-                    copy_chunks!(X.chunks, storeind, B.chunks, ind, l0)
+                    copy_chunks!(Xc, storeind, Bc, ind, l0)
                     storeind += l0
                 end)
         return X
     end
 end
 
-# general multidimensional non-scalar indexing
-
-@generated function unsafe_getindex(B::BitArray, I::Union(Int,AbstractVector{Int})...)
+# in the general multidimensional non-scalar case, can we do about 10% better
+# in most cases by manually hoisting the bitarray chunks access out of the loop
+# (This should really be handled by the compiler or with an immutable BitArray)
+@generated function _unsafe_getindex!(X::BitArray, ::LinearFast, B::BitArray, I::Union(Int,AbstractVector{Int},Colon)...)
```

Comment:
This fits into the general schema I've set up for all `AbstractArrays`, which jumps through the following functions:

* `getindex(::AbstractArray, ::Any...)` - simply call `linearindexing` and redispatch to
* `_getindex(::LinearIndexing, ::AbstractArray, ::Union(Real, AbstractVector, Colon))` - `checkbounds` and call:
* `_unsafe_getindex(::LinearIndexing, ::AbstractArray, ::Union(Real, AbstractVector, Colon))` - create output array, double-check its size, and call `to_index`, and then dispatch to
* `_unsafe_getindex!` (this function).

So, yes, this could be called with an `AbstractVector`, but it's already passed through `to_index` and its element type is restricted to be `Int`.